💥 v2 (#30)
exah/ya-fetch: README.md

```
TODO
```

#### → blob(): Promise\<Blob>

```
TODO
```

#### → void(): Promise\<void>

```
TODO
```

### options

#### resource?: string

Part of the request URL. If used multiple times all the parts will be combined in final URL. The same as first argument of `get`, `post`, `patch`, `put`, `delete`, `head`.

```ts
const instance = YF.create({
  resource: 'https://jsonplaceholder.typicode.com',
})

// will me merged and send request to 'https://jsonplaceholder.typicode.com/posts'
await instance.get('/posts')

// same as
await YF.get('https://jsonplaceholder.typicode.com/posts')

// will me merged to 'https://jsonplaceholder.typicode.com/posts'
const posts = instance.extend({
  resource: '/posts',
})

// will send request to 'https://jsonplaceholder.typicode.com/posts'
await posts.get()
```

#### headers?: HeadersInit

Request headers. Multiple `headers` will merged together.

```ts
const instance = YF.create({
  headers: { 'x-from': 'Website' },
})

// will use instance `headers`
await instance.get('https://jsonplaceholder.typicode.com/posts')

// will be merged with instance `headers`
const authorized = instance.extend({
  headers: { Authorization: 'Bearer token' },
})

> // will send be send with `Authorization` and `x-from` headers

Reviewer: ```suggestion
// will be sent with `Authorization` and `x-from` headers
```

```ts
await instance.patch('/posts/1', { json: { title: 'Hey' } })
```

#### params?: URLSearchParams | object | string

Search params to append to the request URL. Provide an `object`, `string`, or `URLSearchParams` instance. The `object` will be stringified with [`serialize`](#serialize-params-object-urlsearchparams--string) function.

```ts
// request will be sent to 'https://jsonplaceholder.typicode.com/posts?userId=1'
await instance.get('/posts', { params: { userId: 1 } })
```

#### serialize?: (params: object): URLSearchParams | string

Custom search params serializer when `object` is used. Defaults to internal implementation based on `URLSearchParams` with better handling of array values.

```ts
import queryString from 'query-string'

const instance = YF.create({
  resource: 'https://jsonplaceholder.typicode.com',
  serialize: (params) =>
    queryString.stringify(params, {
      arrayFormat: 'bracket',
    }),
})

// request will be sent to 'https://jsonplaceholder.typicode.com/posts?userId=1&tags[]=1&tags[]=2'
await instance.get('/posts', { params: { userId: 1, tags: [1, 2] } })
```

#### timeout?: number

> If specified `TimeoutError` will be thrown and the request will be cancelled after the specified duration.

Reviewer: ```suggestion
If specified, `TimeoutError` will be thrown and the request will be cancelled after the specified duration.
```

```ts
try {
  await instance.get('/posts', { timeout: 500 })
} catch (error) {
  if (error instanceof TimeoutError) {
    // do something, or nothing
  }
}
```

#### onRequest?(options: RequestOptions): Promise\<void> | void

Request handler. Use the callback to modify options before the request or cancel the request. Please, note the options here are in the final state before the request will be made, this means `resource` is string, `params` is instance of `URLSearchParams`, and `headers` is instance of `Headers`.

```ts
const authorized = instance.extend({
  async onRequest(options) {
    // fetch 'token' somehow

>     options.headers.set('Authorization', `Bearer ${await 'token'}`)

Reviewer: are there meant to be quotes around 'token'?

  },
})

> // will include `Authorization` will be resolved with async `Bearer token`.

Reviewer: ```suggestion
// request will be sent with `Authorization` header resolved with async `Bearer token`.
```

await authorized.get('/posts')
```

```ts

> const cancellable = instance.extend({

Reviewer: it could be nice to have an "abort controller support" in the top list linking to this example, wdyt?
(and maybe this example might be moved to options list into signal header section)

```ts
const instance = YF.create({
  onSuccess(response) {
    // you can modify the response in any way you want
    // or even make a new request
    return new Response(response.body, response)
  },
})
```

#### onFailure?(error: ResponseError | TimeoutError | Error): Promise\<Response> | Response

Throw custom error with additional data, return a new `Promise` with `Response` using `request`, or just submit an event to error tracking service.

```ts
class CustomResponseError extends YF.ResponseError {
  data: unknown
  constructor(response: YF.Response, data: unknown) {
    super(response)
    this.data = data
  }
}

const api = YF.create({
  resource: 'http://localhost',
  async onFailure(error) {
    if (error instanceof YF.ResponseError) {
      if (error.response.status < 500) {
        throw new CustomResponseError(
          error.response,
          await error.response.json()
        )
      }
    }

    trackError(error)
    throw error
  },
})
```

### request

```
TODO
```

## 🔗 Alternatives
### ResponseError

- [`ky`](https://github.com/sindresorhus/ky) - Library that inspired this one, but 3x times bigger and not transpiled for es5 browsers

> - [`axios`](https://github.com/axios/axios) - Based on old `XMLHttpRequests` API, almost 5x times bigger, but super popular and feature packed

Reviewer: :)
